[RNMobile] Remove third-party dependency on react-native-hsv-color-picker

derekblank opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

Related:

* https://github.com/wordpress-mobile/gutenberg-mobile/issues/5847
* https://github.com/wordpress-mobile/gutenberg-mobile/pull/6029

Todo:
- [x] Add integration tests to Color Picker
- [x] Migrate tinycolor2 references to use colord library


## What?
Replaces the forked version of [react-native-hsv-color-picker](https://github.com/wordpress-mobile/react-native-hsv-color-picker/) in favor of implementing the library within Gutenberg instead.

## Why?
To reduce the number of third-party libraries and dependencies in order to better support React Native's [New Architecture](https://reactnative.dev/docs/new-architecture-intro) and simplify the React Native upgrade process for Gutenberg Mobile.

## How?
The PR takes the source of our forked react-native-hsv-color-picker code, and reimplements the source within the Color Picker component. The new HSV components (`HuePicker`, `SaturationValuePicker`) are just UI interfaces that leverage the libraries `tinycolor2` and `react-native-linear-gradient`, which are conveniently already dependencies of the project.

## Testing Instructions
1. Create a post
2. Add a block that implements the color picker (like Cover block, and select the gradient option)
3. Observe that the Color Picker behavior has not changed

## Screenshots or screencast <!-- if applicable -->

<video src="https://github.com/WordPress/gutenberg/assets/643285/1601ef5c-13ec-41f9-92c2-8e01a83b21ac" />

derekblank committed:
Add hsv color picker components

derekblank committed:
Update Saturation Picker from source

derekblank committed:
Update Hue Picker from source

derekblank committed:
Update hsv-color-picker imports

derekblank committed:
Remove third-party dependency package react-native-hsv-color-picker

derekblank committed:
Update hsv-color-picker styles

derekblank committed:
Merge remote-tracking branch 'origin' into rnmobile/remove-react-native-hsv-color-picker

derekblank committed:
Update CHANGELOG

derekblank committed:
Add tinycolor2 package

derekblank committed:
Remove dependency on tinycolor2 from hsv-color-picker

derekblank committed:
Remove tinycolor package reference

derekblank committed:
Revert empty change on initial editor html

derekblank committed:
Merge trunk

derekblank committed:
Update hsv-color-picker container style

derekblank committed:
Add color-picker integration test

derekblank commented:
@dcalhoun Let me know if you have any thoughts on the testing strategy here. Initially, I had planned to add tests to the ColorPicker component simply because it didn't have tests. While writing the tests, I realized the ColorPicker functionality was mostly covered by tests that already existed, like [this Cover block test](https://github.com/WordPress/gutenberg/blob/0a7317728ffa6d342359ea384c5bce567b7db6e1/packages/block-library/src/cover/test/edit.native.js#L543-L575), so this test may not be needed. At any rate, I don't think I've mocked the test setup correctly -- I was trying to avoid initializing the full editor by just testing the ColorPicker itself. If we do decide to keep this test, perhaps it's better to initialize the full editor here?

dcalhoun commented on package-lock.json:
The merge conflict for this file may relate to #53426. It is likely we merely need to pull in the latest `trunk` branch and `npm run distclean && npm run install`.

dcalhoun commented on packages/block-library/src/cover/test/edit.native.js:
```diff
@@ -541,7 +540,7 @@ describe( 'color settings', () => {
 
 	it( 'displays the hex color value in the custom color picker', async () => {
-		HsvColorPicker.mockImplementation( ( props ) => {
+		ColorPicker.mockImplementation( ( props ) => {
```
This test fails currently as `ColorPicker` is not mocked as `HsvColorPicker` was [and still is](https://github.com/WordPress/gutenberg/blob/20ae378b7065bf75b2af6809b179dfa38028eb4b/test/native/setup.js#L191-L195) — we should remove that outdated `HsvColorPicker` mock now that it is no longer a part of the project. 

I originally mocked `HsvColorPicker` as it is a third-party dependency, one with its own tests. Now that this is an internal dependency, it is likely we should no longer mock this, testing the functionality of the new code.

derekblank replied:
In context of [this comment](https://github.com/WordPress/gutenberg/pull/53329#discussion_r1291507353), I've updated this Cover block test to remove the outdated `HsvColorPicker` mock. However, in the process of digesting the (now removed) Color Picker test in favor of this one, I think I've confused myself on the best place to implement the `testID="color-picker"` param that is being queried [here](https://github.com/WordPress/gutenberg/blob/20ae378b7065bf75b2af6809b179dfa38028eb4b/packages/block-library/src/cover/test/edit.native.js#L564). In best practice, should this be added to the [Color Picker params itself](https://github.com/WordPress/gutenberg/blob/59a346da13c8c0d759bfe9609c15afb737922e53/packages/components/src/color-picker/index.native.js#L123-L157), or to the [Cover block implementation](https://github.com/WordPress/gutenberg/blob/59a346da13c8c0d759bfe9609c15afb737922e53/packages/block-library/src/cover/edit.native.js#L391-L411) of the Color Picker?

dcalhoun replied:
My high level answer would be to keep it simple. Place the test ID directly on the element you need to query (the element handling the event callback), rather than attempting to pass it down from a parent or test context. 

Passing down could make the the components more flexible, but avoid that complexity until we need it.

derekblank replied:
Thanks. I ended up placing it on the [hsv-color-picker component](https://github.com/WordPress/gutenberg/pull/53329/commits/d14b5f62af9c4d74e6c3e04be5962abac1fbc510) itself.

dcalhoun commented on packages/components/src/color-picker/hue-picker.native.js:
```diff
@@ -0,0 +1,194 @@
+/**
+ * Internal dependencies
+ */
+import LinearGradient from 'react-native-linear-gradient';
```
We should probably group this external dependencies with the rest atop this file.

dcalhoun commented on packages/components/src/color-picker/hue-picker.native.js:
```diff
@@ -0,0 +1,194 @@
+ * Internal dependencies
+ */
+import LinearGradient from 'react-native-linear-gradient';
+import styles from './style.native.scss';
```
We may be able to drop the `native` reference and rely upon Metro's platform-specific module resolution. I could be wrong. 
```suggestion
import styles from './style.scss';
```

dcalhoun commented on packages/components/src/color-picker/hue-picker.native.js:
```diff
@@ -0,0 +1,194 @@
+import LinearGradient from 'react-native-linear-gradient';
+import styles from './style.native.scss';
+
+export default class HuePicker extends Component {
```
> Convert the code from a class to a function component in the process, and modify any library code that would better suit the project.

Given the above referenced goal from https://github.com/wordpress-mobile/gutenberg-mobile/issues/5847, is the retention of some classes done for the sake of time?

derekblank replied:
> Given the above referenced goal from https://github.com/wordpress-mobile/gutenberg-mobile/issues/5847, is the retention of some classes done for the sake of time?

Yes, absolutely. I did explore migrating the Saturation and Hue picker to a function component as stated in the goal, but ran into complications with the [Saturation picker ref](https://github.com/WordPress/gutenberg/pull/53329/files#diff-53f55d9d33d9322806946cf398f41f3febce3f5c21b6a4402194181d69378e87R66) that would likely have required forwarding the ref and implementing [useImperativeHandle](https://react.dev/reference/react/useImperativeHandle). Given the relatively low priority of removing this library, I timeboxed these explorations and was hesitant to modify too much of the existing code to avoid regressions. 

That being said, if you note a clearer path forward for migrating these to function components that I may not be considering, feel free to suggest, and we can increment.

dcalhoun commented on packages/components/src/color-picker/index.native.js:
```diff
@@ -122,6 +122,7 @@ function ColorPicker( {
 		<>
 			<HsvColorPicker
 				huePickerHue={ hue }
+				currentColor={ currentColor }
```
Is the addition of this prop and removal of [`tinycolor` usage](https://github.com/lukewalczak/react-native-hsv-color-picker/blob/163e78cd4d986a28954a626213b40daee15774e8/src/HuePicker.js#L84C1-L87) a simplification of the component to fit our needs?

derekblank replied:
It was brought to my attention that there was already a [previous effort](https://github.com/WordPress/gutenberg/pull/35562) to remove the tinycolor package from Gutenberg in favor of using [colord](https://github.com/omgovich/colord), thus I refactored the tinycolor references to use the colord API. Adding the `currentColor` prop here is (what seems like) an efficient use of prop-drilling, as I noticed the HSV-to-hex value is already being calculated [here](https://github.com/WordPress/gutenberg/blob/20ae378b7065bf75b2af6809b179dfa38028eb4b/packages/components/src/color-picker/index.native.js#L79-L82), which is being utilized to set the background color of the Saturation picker [here](https://github.com/WordPress/gutenberg/blob/20ae378b7065bf75b2af6809b179dfa38028eb4b/packages/components/src/color-picker/saturation-picker.native.js#L152). Let me know if you are noting a more efficient way to pass this combined hex color to the child components, however.

dcalhoun commented on packages/components/src/color-picker/style.native.scss:
```diff
@@ -62,3 +62,26 @@
+	top: 0;
+	left: 0;
+	position: absolute;
+	border-color: #fff;
```
Likely doesn't matter at all, as I do not believe anyone is ever going to change the definition of "white," but should we rely upon Sass variables for `#fff`? There is one other instance in this file as well. 
```suggestion
	border-color: $white;
```

dcalhoun commented on packages/react-native-editor/CHANGELOG.md:
```diff
@@ -11,6 +11,7 @@ For each user feature we should also add a importance categorization label  to i
 
 ## Unreleased
 -   [*] Display custom color value in mobile Cover Block color picker [#51414]
+-   [**] Replace third party dependency on react-native-hsv-color-picker
```
Expanded this for clarity, but feel free to adjust or ignore as you see fit. 
```suggestion
-   [**] Replace third-party dependency react-native-hsv-color-picker with first-party code [#53329]
```

dcalhoun commented on packages/components/src/color-picker/test/index.native.js:
Referencing https://github.com/WordPress/gutenberg/pull/53329#issuecomment-1674276338 here as it largely pertains to this file. 

> [...] At any rate, I don't think I've mocked the test setup correctly -- I was trying to avoid initializing the full editor by just testing the ColorPicker itself. If we do decide to keep this test, perhaps it's better to initialize the full editor here?

Testing with a "scoped" approach (rendering the minimum) rather than initializing the full editor is completely valid IMO. It is a judgement call of trade-offs, e.g. test speed/complexity vs test robustness. 

The biggest issue with the current test failure is that the [queried test ID](https://github.com/WordPress/gutenberg/blob/20ae378b7065bf75b2af6809b179dfa38028eb4b/packages/components/src/color-picker/test/index.native.js#L22) does not exist anywhere. This is due to the `ColorPicker` component not passing the receive `testID` prop to any receiving native element. If we were to keep this test, I might address the failures with the following changes. 

<details><summary>Test changes diff</summary>

```diff
diff --git a/packages/components/src/color-picker/hue-picker.native.js b/packages/components/src/color-picker/hue-picker.native.js
index 570dd37396..cec9849e64 100644
--- a/packages/components/src/color-picker/hue-picker.native.js
+++ b/packages/components/src/color-picker/hue-picker.native.js
@@ -155,6 +155,7 @@ export default class HuePicker extends Component {
 					left: 0,
 					right: 0,
 				} }
+				testID="hue-picker"
 			>
 				<LinearGradient
 					colors={ hueColors }
diff --git a/packages/components/src/color-picker/test/index.native.js b/packages/components/src/color-picker/test/index.native.js
index dc3f9c21ba..7d567e1893 100644
--- a/packages/components/src/color-picker/test/index.native.js
+++ b/packages/components/src/color-picker/test/index.native.js
@@ -1,41 +1,41 @@
 /**
  * External dependencies
  */
-import { render, fireEvent } from 'test/helpers';
+import { screen, render, fireEvent } from 'test/helpers';
 
 /**
- * WordPress dependencies
+ * Internal dependencies
  */
-import { ColorPicker } from '@wordpress/components';
+import { ColorPicker } from '../index';
 
 describe( 'color picker', () => {
 	it( 'selects the correct color', async () => {
 		const shouldEnableBottomSheetMaxHeight = jest.fn();
 		const onHandleClosingBottomSheet = jest.fn();
+		const setColor = jest.fn();
 
-		const { getByTestId } = render(
+		render(
 			<ColorPicker
+				bottomLabelText="Select a color"
 				shouldEnableBottomSheetMaxHeight={
 					shouldEnableBottomSheetMaxHeight
 				}
 				onHandleClosingBottomSheet={ onHandleClosingBottomSheet }
-				testID="color-picker"
+				setColor={ setColor }
 			/>
 		);
 
-		const colorPicker = getByTestId( 'color-picker' );
-
 		// Assert label text before tapping color picker
-		expect( colorPicker.getByText( 'Select a color' ) ).toBeVisible();
+		expect( screen.getByText( 'Select a color' ) ).toBeVisible();
 
 		// Tap color picker
-		fireEvent( colorPicker, 'onHuePickerPress', {
+		fireEvent( screen.getByTestId( 'hue-picker' ), 'onPress', {
 			hue: 120,
 			saturation: 12,
 			value: 50,
 		} );
 
 		// Assert label hex value after tapping color picker
-		expect( colorPicker.getByText( '#00FF00' ) ).toBeVisible();
+		expect( screen.getByText( '#00FF00' ) ).toBeVisible();
 	} );
 } );

```

</details> 

However, even with these changes, it may not be appropriate to assert this feature at this level. The HEX value label is controlled _above_ `ColorPicker` via `bottomLabelText` within the [Cover block edit component](https://github.com/WordPress/gutenberg/blob/20ae378b7065bf75b2af6809b179dfa38028eb4b/packages/block-library/src/cover/edit.native.js#L373-L379). The only way to assert this feature at this level would be to (A) remove `bottomLabelText` via `update(<ColorPicker bottomLabelText={null} />` or (B) load the entire editor, add a Cover block, and test it. The latter may be misguided, as it might imply this feature is managed by `ColorPicker` alone.

derekblank replied:
> The only way to assert this feature at this level would be to (A) remove bottomLabelText via update(<ColorPicker bottomLabelText={null} /> or (B) load the entire editor, add a Cover block, and test it. The latter may be misguided, as it might imply this feature is managed by ColorPicker alone.

I agree. Implementing a parent block (like Cover) was what I was trying to avoid here. After applying your patch and reviewing the test strategy options, I think the assertion we're after is already better covered by the Cover block. I've removed this test, but thank you for your detailed input.

dcalhoun reviewed (commented):
Nice work! Thank you for helping address this complexity. 

I'd like to test the implementation on physical devices and review the larger files more thoroughly, but I wanted to share initial feedback now. I'll follow up later. 

---

> @dcalhoun Let me know if you have any thoughts on the testing strategy here. Initially, I had planned to add tests to the ColorPicker component simply because it didn't have tests. While writing the tests, I realized the ColorPicker functionality was mostly covered by tests that already existed, like [this Cover block test](https://github.com/WordPress/gutenberg/blob/0a7317728ffa6d342359ea384c5bce567b7db6e1/packages/block-library/src/cover/test/edit.native.js#L543-L575), so this test may not be needed.

Good catch. I might describe this decision as unit testing (`ColorPicker`) vs integration testing (`ColorPicker` via Cover edit). I think there is an argument for either approach. I also think there is a balance as to how many tests we have, i.e. should we test _both_ the unit (`ColorPicker` alone) and integration (`ColorPicker` within Cover and Buttons)? 

Most of our current tests target edit components, rather than specific units within them. **So, my thought is testing via Cover and Buttons edit makes sense; keeping in mind that we may should consider avoiding writing identical test cases in both locations.** 

That said, `ColorPicker` is a complex unit, used a in a few different places. If we want to unit test it, I could understand that. 

Lastly, we also have a pattern of testing subjects that cut across numerous locations. E.g. [editor history tests](https://github.com/WordPress/gutenberg/blob/69d9ca3b9524334af5c8936179131f02d2208a9d/test/native/integration/editor-history.native.js) are placed in a top-level location as it relates to several different edit components. My thought is that color picking doesn't reach this level of ubiquity, but I'm open to thoughts on that.

derekblank committed:
Merge trunk

derekblank committed:
Update packages/components/src/color-picker/style.native.scss

Co-authored-by: David Calhoun <github@davidcalhoun.me>

derekblank committed:
Update packages/react-native-editor/CHANGELOG.md

Co-authored-by: David Calhoun <github@davidcalhoun.me>

derekblank committed:
Update Color Picker code style and test formatting

derekblank committed:
Merge branch 'rnmobile/remove-react-native-hsv-color-picker' of https://github.com/WordPress/gutenberg into rnmobile/remove-react-native-hsv-color-picker

derekblank commented:
@dcalhoun Thanks for the detailed review! I've added some comments, with the only question I had on [this comment](https://github.com/WordPress/gutenberg/pull/53329#discussion_r1292934690). The rest of the feedback I either implemented or agreed with. 👍

derekblank committed:
Update hsv-color-picker test ID

derekblank committed:
Merge branch 'trunk' into rnmobile/remove-react-native-hsv-color-picker

derekblank committed:
Remove color picker test

Removes Color Picker test in favor of testing the color picker
in a deeper editor tree, like Cover block

derekblank committed:
Update saturation picker imports

derekblank commented:
@dcalhoun I think this one should be ready to review again. Let me know if you have further feedback! 🙇

dcalhoun reviewed (approved):
This looks good! Thank you for completing this. I tested the performance using prototype builds and did not note any degradations.

derekblank committed:
Merge branch 'trunk' into rnmobile/remove-react-native-hsv-color-picker
